### PR TITLE
Blog for Log Forwarding API

### PR DESCRIPTION
**Pull request by halim-lee**

### CODE REVIEW

**Comment by halim-lee** [comment]
Blog post preview available here: http://elf1.fyre.ibm.com:4000/blog/2020/08/16/ocp-log-forwarding.html

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 1]
```diff
@@ -0,0 +1,224 @@
+---
+layout: post
+title: "Forwarding Open Liberty logs to external cluster using OCP Log Forwarding API"
+categories: blog
+author_picture: https://avatars3.githubusercontent.com/leemelim
+author_github: https://github.com/leemelim
+seo-title: Forwarding Open Liberty logs to external cluster using OCP Log Forwarding API - OpenLiberty.io
+seo-description: Using Log Forwarding API, you can send logs from Open Liberty deployed in OpenShift Container Platform to remote destinations. Logs, which get collected by cluster logging in OpenShift Container Platform, can be forwarded to either external Elasticsearch cluster or external log aggregation solution using forward protocol.
+blog_description: Using Log Forwarding API, you can send logs from Open Liberty deployed in OpenShift Container Platform to remote destinations. Logs, which get collected by cluster logging in OpenShift Container Platform, can be forwarded to either external Elasticsearch cluster or external log aggregation solution using forward protocol.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: - would it make sense to call it the Fluentd forward protocol?  not sure people will know what "forward" protocol is on its own

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 2]
```diff
@@ -0,0 +1,224 @@
+---
+layout: post
+title: "Forwarding Open Liberty logs to external cluster using OCP Log Forwarding API"
+categories: blog
+author_picture: https://avatars3.githubusercontent.com/leemelim
+author_github: https://github.com/leemelim
+seo-title: Forwarding Open Liberty logs to external cluster using OCP Log Forwarding API - OpenLiberty.io
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: in a few places you refer simply to "external cluster".  it makes me wonder what it's an external cluster of.

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+blog_description: Using Log Forwarding API, you can send logs from Open Liberty deployed in OpenShift Container Platform to remote destinations. Logs, which get collected by cluster logging in OpenShift Container Platform, can be forwarded to either external Elasticsearch cluster or external log aggregation solution using forward protocol.
+---
+= Forwarding Open Liberty logs to external cluster using OCP Log Forwarding API
+Halim Lee <https://github.com/leemelim>
+
+Container orchestration system is highly distributed and its parts are dynamic. It becomes difficult to detect which worker nodes are running your containers and pods for your application. Application logging and log collection helps you monitor and troubleshoot these applications efficiently. Openshift Container Platform provides a log aggregation solution with cluster logging. The cluster logging is consisted of three components: Elasticsearch, Fluentd, and Kibana (EFK). Fluentd is deployed to each node in the OpenShift Container Platform cluster and forwards all logs to Elasticsearch. Elasticsearch stores and organizes the log data. Kibana helps you create visualizations and searches with the aggregated data.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: Openshift -> OpenShift

Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: Should write an intro paragraph that relates to the problem that you think your users are having.  Maybe they are new to OpenShift and while they can see it has built in Elasticsearch support they are already using Splunk to monitor apps outside of their k8s environment and they'd like to send all their logs to one place to make it easier to manage and to see possible event correlations.  They might wish they could hijack the Fluentd in cluster logging to send logs externally instead...

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+----
+[root@ocp ~]# oc set env ds/fluentd MERGE_JSON_LOG=true
+----
++
+
+. To check if the logs are being forwarded to the specified outputs, run the following command:
++
+[source]
+----
+[root@ocp ~]# oc -n openshift-logging get cm fluentd -o json | jq -r '.data."fluent.conf"' > fluentd-with-logfowarding.conf
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc:173]: would be good to explain this command a bit so reader knows what it's doing

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+      @label @FLUENTD-FORWARD
+    </store>
+  </match>
+</label>
+...
+```
++
+
+
+== Splunk configuration on external cluster
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: should probably add some intro in this section to say that the idea is that you're going to set up splunk in an external openshift cluster.  Also good to give credit to the person whose git repo you're sending people off to use for the splunk setup.

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+```
++
+
+. Go to Search & Reporting and search for `index="openshift"` to view logs from OpenShift Container Platform.
+
+. Download Splunk dashboards for Open Liberty: link:https://github.com/WASdev/sample.dashboards/tree/master/Liberty/Splunk%208[Sample dashboard for Liberty using Splunk]. Import downloaded sample dashboards using *Source* option. Using this dashboard, you can visualize message, trace, and first failure data capture (FFDC) logging data collected from JSON logging in Open Liberty.
+
+image::/img/blog/splunk-dashboard.png[Splunk-Dashboard,width=70%,align="center"]
+
+== External Elasticsearch and Kibana configuration
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: Good here to also introduce what this setup is going to give you.  in this case, looks like you're going to stand up elasticsearch and kibana using docker compose.  say why you chose that, even if it's just a handy way to set it up.

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+[root@ocp ~]# docker-compose up
+----
++
+
+. Download Kibana dashboards for Open Liberty: link:https://github.com/OpenLiberty/open-liberty-operator/tree/master/deploy/dashboards/logging[Sample Kibana dashboards for Open Liberty]. Import downloaded sample dashboards in Kibana Management > Saved Objects.
+
+image::/img/blog/kibana-dashboard.png[Kibana Dashboard,width=70%,align="center"]
+
+== Conclusion
+Application logging is one of the fundamental types of observability. It helps you monitor and find the problems on your servers easily. Using Log Forwarding API, you can use existing external enterprise log collection solutions for OpenShift Container Platform logs. We have now seen two popular log collection solutions: Splunk and Kibana. Both applications allow you to aggregate and analyze your Open Liberty log events on OpenShift Cluster Platform and determine how healthy and well-performing your Liberty servers are.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: Both applications allow you to aggregate and analyze your Open Liberty log events on OpenShift Cluster Platform
->
Both log analysis solutions allow you to aggregate and analyze log events from Open Liberty servers running on OpenShift Cluster Platform...

**Review by donbourne — CHANGES_REQUESTED** [review]

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 8]
```diff
@@ -1,17 +1,17 @@
...
+Container orchestration system is highly distributed and its parts are dynamic. It becomes difficult to detect which worker nodes are running your containers and pods for your application. Application logging and log collection help you monitor and troubleshoot these applications efficiently. OpenShift Container Platform provides a log aggregation solution with cluster logging, which is consisted of three components: Elasticsearch, Fluentd, and Kibana (EFK). Despite the log aggregation solution provided within OpenShift Cluster Platform, forwarding logs to external locations may be essential. You may already have exisiting log analysis solutions to monitor applications outside of OpenShift Cluster. To aggregate all logs, both logs inside and outside of your OpenShift cluster, you would want to ship the logs generated within OpenShift environment to your exisiting solutions.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: First sentence needs to draw me in.  The usual refrains of "distributed systems make debugging hard" are somewhat overdone, so you need to clearly start with what it is someone is really going to want to read this blog for -- in the first sentence.

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+---
+= Forwarding Open Liberty logs in OpenShift Container Platform to external logging solutions using Log Forwarding API
+Halim Lee <https://github.com/leemelim>
+
+Container orchestration system is highly distributed and its parts are dynamic. It becomes difficult to detect which worker nodes are running your containers and pods for your application. Application logging and log collection help you monitor and troubleshoot these applications efficiently. OpenShift Container Platform provides a log aggregation solution with cluster logging, which is consisted of three components: Elasticsearch, Fluentd, and Kibana (EFK). Despite the log aggregation solution provided within OpenShift Cluster Platform, forwarding logs to external locations may be essential. You may already have exisiting log analysis solutions to monitor applications outside of OpenShift Cluster. To aggregate all logs, both logs inside and outside of your OpenShift cluster, you would want to ship the logs generated within OpenShift environment to your exisiting solutions.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: need to mention, near the beginning, that the Log Fowarding API is a tech preview.

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+  </match>
+</label>
+...
+```
++
+
+
+== Internal and External Splunk configuration
+
+Using `forward` output, you can forward OpenShift Container Platform logs to Splunk using Fluentd forward protocol. You can setup Splunk inside your OpenShift Cluster or on your external machine.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: as discussed, refactor this section to make the external splunk scenario be your primary one and refer to the RH article for the splunk in OpenShift one.

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
++
+[source]
+----
+[root@ocp ~]# oc get pods -n openshift-logging
+----
++
+
+=== Option 2: Setting up Splunk on your external machine
+
+This option sets up Splunk manually on your external machine. If you already have Splunk deployed on your external machine, this option will help you setup the connection between your OpenShift cluster and Splunk. You can deploy Fluentd and Splunk using docker compose. This way, obtaining docker images, setting up configurations and running the applications can be done easily.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: should mention here that this docker compose way of doing setup is really for demo purposes

Comment (donbourne) [reply posts/2020-08-16-ocp-log-forwarding.adoc]: also would be good to educate reader on the fact that Fluentd forward forwards to another Fluentd instance, which is why you need it in front of your external Splunk

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+=== Setting up Splunk Dashboard
+
+. Go to Search & Reporting. Search for `index="openshift"` for *Option 1* and `source="openshift"` for *Option 2* to view logs from OpenShift Container Platform.
+
+. Download Splunk dashboards for Open Liberty: link:https://github.com/WASdev/sample.dashboards/tree/master/Liberty/Splunk%208[Sample dashboard for Liberty using Splunk]. Import downloaded sample dashboards using *Source* option. Using this dashboard, you can visualize message, trace, and first failure data capture (FFDC) logging data collected from JSON logging in Open Liberty.
+
+image::/img/blog/splunk-dashboard.png[Splunk-Dashboard,width=70%,align="center"]
+
+== External Elasticsearch and Kibana configuration
+Using `elasticsearch` output, you can forward logs to Elasticsearch cluster deployed on your external machine. You run Elasticsearch and Kibana with docker compose. This way, obtaining docker images, setting up configurations and bringing up the clusters can be done  using one YAML file.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: mention that the docker compose approach to setting up elasticsearch and kibana is just for demo purposes

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+[root@ocp ~]# docker-compose up
+----
++
+
+. Download Kibana dashboards for Open Liberty: link:https://github.com/OpenLiberty/open-liberty-operator/tree/master/deploy/dashboards/logging[Sample Kibana dashboards for Open Liberty]. Import downloaded sample dashboards in Kibana Management > Saved Objects.
+
+image::/img/blog/kibana-dashboard.png[Kibana Dashboard,width=70%,align="center"]
+
+== Conclusion
+Application logging is one of the fundamental types of observability. It helps you monitor and find the problems on your servers easily. Using Log Forwarding API, you can use existing external enterprise log collection solutions for OpenShift Container Platform logs. We have now seen two popular log collection solutions: Splunk and Kibana. Both log analysis solutions allow you to aggregate and analyze log events from Open Liberty servers running on OpenShift Cluster Platform.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: "it helps you monitor" - the term "monitor" is generally used in this space to talk about gathering the metrics or logs from a server, so may be confusing the way you're using it here.

Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc]: We have now seen two popular log collection solutions: Splunk and Kibana. Both log analysis solutions allow you to aggregate and analyze log events from Open Liberty servers running on OpenShift Cluster Platform. 
->
We have now seen two popular log collection solutions - one based on Fluentd / Splunk and the other based on Fluentd / Elasticsearch / Kibana. Both log analysis solutions allow you to aggregate and analyze log events from Open Liberty servers running on OpenShift Cluster Platform.

**Review by donbourne — CHANGES_REQUESTED** [review]

**Review by donbourne — DISMISSED** [review]

**Review by donbourne — DISMISSED** [review]

**Comment by donbourne** [comment]
plan is to do a test run through this before publishing.

**Comment by donbourne** [comment]
remember to merge PR https://github.com/WASdev/sample.dashboards/pull/18 after this blog is published (the corresponding splunk dashboard)

**Review by donbourne — DISMISSED** [review]

**Comment by Channyboy** [comment]
@lauracowen 
Hi Laura, there have been updates made to this blog.
May you review the new updates.

**Comment by lauracowen** [comment]
@Channyboy Can you get someone in your team to review it and then merge/build it on draft? We'll review/edit it after that. Anyone can review it so that you can merge. Thanks

**Review by fmhwong — DISMISSED** [review]
`log-forwarding.png` doesn't seem to be used in the doc.

**Review by fmhwong — APPROVED** [review]
LGTM

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+seo-description: Using Log Forwarding API, you can send logs from Open Liberty deployed in OpenShift Container Platform to remote destinations. You can deploy Splunk on your external machine or inside your OpenShift Container Platform. The logs can be forwarded to Splunk using the Fluentd forward protocol.
+blog_description: Using Log Forwarding API, you can send logs from Open Liberty deployed in OpenShift Container Platform to remote destinations. You can deploy Splunk on your external machine or inside your OpenShift Container Platform. The logs can be forwarded to Splunk using the Fluentd forward protocol.
+---
+= Forwarding Open Liberty logs in OpenShift Container Platform to Splunk using Log Forwarding API
+Halim Lee <https://github.com/leemelim>
+
+OpenShift Container Platform provides a log aggregation solution with cluster logging using Elasticsearch, Fluentd, and Kibana (EFK stack). In addition to EFK stack, integration with other logging applications has been made available. Starting OpenShift 4.3, you can use Log Forwarding API, which is currently in Technology preview, to send logs from Open Liberty in OpenShift cluster to other available log analysis solutions.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc:14]: In addition to EFK stack, integration with other logging applications has been made available
->
In addition to EFK stack, you can integrate with other log analysis solutions

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 15]
```diff
@@ -0,0 +1,421 @@
...
+
+Splunk is one of the most popular logging solutions. If you already have Splunk deployed on your external machine, or if you want to start using Splunk as your new logging solution, you will have to integrate OpenShift with Splunk. Collecting logs in one location helps you monitor all of your applications efficiently, especially because container orchestration system is highly distributed and dynamic. Log Forwarding API simplifies the integration ability of OpenShift cluster with Splunk. You can configure custom pipelines to send logs to desired endpoints within or outside of your OpenShift cluster.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc:16]: Splunk is one of the most popular logging solutions. 
->
Splunk is one of the most popular log analysis solutions.

File: posts/2020-08-16-ocp-log-forwarding.adoc [hunk 16]
```diff
@@ -0,0 +1,421 @@
...
+
+
+== Configuring Log Forwarding API
+
+. Ensure your cluster logging instance is created and all pods are fully operational. See the installation and configuration guide for cluster logging: link:https://docs.openshift.com/container-platform/4.5/logging/cluster-logging-deploying.html[Deploying cluster logging]. It should be noted that the provided link leads to the OCP 4.5 documentation. You can chose the appropriate version once on the site. Additionally, the cluster logging instance sample provided on the RedHat documentation leaves the _storage class_ value for the reader to configure. If you are simply _testing_ you can omit the storage section and the cluster logging instance will use in-memory storage. Note that this configuration would only be suitable for testing and NOT for production.
```
Comment (donbourne) [inline posts/2020-08-16-ocp-log-forwarding.adoc:25]: RedHat 
-> 
Red Hat

**Pull request merged**